File: bumperbot_controller/bumperbot_controller/noisy_controller.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
from sensor_msgs.msg import JointState
from nav_msgs.msg import Odometry
import numpy as np
from rclpy.time import Time
from rclpy.constants import S_TO_NS
import math
from tf_transformations import quaternion_from_euler
from tf2_ros import TransformBroadcaster


class NoisyController(Node):

    def __init__(self):
        super().__init__("noisy_controller")

        self.declare_parameter("wheel_radius", 0.033)
        self.declare_parameter("wheel_separation", 0.17)

        self.wheel_radius_ = self.get_parameter(
            "wheel_radius").get_parameter_value().double_value
        self.wheel_separation_ = self.get_parameter(
            "wheel_separation").get_parameter_value().double_value

        # debug
        self.get_logger().info(f"wheel radius: {self.wheel_radius_}")
        self.get_logger().info(f"wheel separation: {self.wheel_separation_}")

        self.left_wheel_prev_pos_ = 0.0
        self.right_wheel_prev_pos_ = 0.0
        self.prev_time_ = self.get_clock().now()

        self.x_ = 0.0
        self.y_ = 0.0
        self.theta_ = 0.0

        
        self.joint_sub_ = self.create_subscription(JointState, "joint_states", self.joint_state_callback, 10)
        
        # odom publisher
        self.odom_pub_ = self.create_publisher(Odometry, "bumperbot_controller/odom_noisy", 10)
        
        self.odom_msg_ = Odometry()
        self.odom_msg_.header.frame_id = "odom"
        self.odom_msg_.child_frame_id = "base_footprint_ekf"
        self.odom_msg_.pose.pose.orientation.x = 0.0
        self.odom_msg_.pose.pose.orientation.y = 0.0
        self.odom_msg_.pose.pose.orientation.z = 0.0
        self.odom_msg_.pose.pose.orientation.w = 1.0

        # tf2 stuff
        self.br_ = TransformBroadcaster(self)
        self.odom_to_base_footprint_tf = TransformStamped()
        self.odom_to_base_footprint_tf.header.frame_id = "odom"
        self.odom_to_base_footprint_tf.child_frame_id = "base_footprint_noisy"

    

    def joint_state_callback(self, msg):
        wheel_encoder_left = msg.position[1] + np.random.normal(0, 0.005)
        wheel_encoder_right = msg.position[0] + np.random.normal(0, 0.005)

        dp_left = wheel_encoder_left - self.left_wheel_prev_pos_
        dp_right = wheel_encoder_right - self.right_wheel_prev_pos_
        dt = Time.from_msg(msg.header.stamp) - self.prev_time_

        # update
        self.left_wheel_prev_pos_ = msg.position[1]
        self.right_wheel_prev_pos_ = msg.position[0]
        self.prev_time_ = Time.from_msg(msg.header.stamp)

        # inverse kinematics
        fi_left = dp_left / (dt.nanoseconds/S_TO_NS)
        fi_right = dp_right / (dt.nanoseconds/S_TO_NS)

        # velocities
        linear_vel = (self.wheel_radius_ * fi_left + self.wheel_radius_ * fi_right) / 2.0
        angular_vel = (self.wheel_radius_ * fi_right - self.wheel_radius_ * fi_left) / self.wheel_separation_

        # position and orientation
        d_s = (self.wheel_radius_ * dp_right + self.wheel_radius_ * dp_left) / 2.0
        d_theta = (self.wheel_radius_ * dp_right - self.wheel_radius_ * dp_left) / self.wheel_separation_
        self.theta_ += d_theta
        self.x_ += (d_s * math.cos(self.theta_))
        self.y_ += (d_s * math.sin(self.theta_))

        q = quaternion_from_euler(0, 0, self.theta_)
        self.odom_msg_.pose.pose.orientation.x = q[0]
        self.odom_msg_.pose.pose.orientation.y = q[1]
        self.odom_msg_.pose.pose.orientation.z = q[2]
        self.odom_msg_.pose.pose.orientation.w = q[3]
        self.odom_msg_.header.stamp = self.get_clock().now().to_msg()
        self.odom_msg_.pose.pose.position.x = self.x_
        self.odom_msg_.pose.pose.position.y = self.y_
        self.odom_msg_.twist.twist.linear.x = linear_vel
        self.odom_msg_.twist.twist.angular.z = angular_vel

        self.odom_to_base_footprint_tf.transform.translation.x = self.x_
        self.odom_to_base_footprint_tf.transform.translation.y = self.y_
        self.odom_to_base_footprint_tf.transform.rotation.x = q[0]
        self.odom_to_base_footprint_tf.transform.rotation.y = q[1]
        self.odom_to_base_footprint_tf.transform.rotation.z = q[2]
        self.odom_to_base_footprint_tf.transform.rotation.w = q[3]
        self.odom_to_base_footprint_tf.header.stamp = self.get_clock().now().to_msg()



        self.odom_pub_.publish(self.odom_msg_)
        self.br_.sendTransform(self.odom_to_base_footprint_tf)

        self.get_logger().info(f"linear velocity of the robot: {linear_vel}, and the angular velocity: {angular_vel}")
        self.get_logger().info(f"x position: {self.x_}, y position: {self.y_}, and theta: {self.theta_}")


def main():
    rclpy.init()
    noisy_controller_node = NoisyController()
    rclpy.spin(noisy_controller_node)
    noisy_controller_node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: bumperbot_controller/src/noisy_controller.cpp
#include "bumperbot_controller/noisy_controller.hpp"
#include <tf2/LinearMath/Quaternion.h>
#include <random>

using std::placeholders::_1;

NoisyController::NoisyController(const std::string &name)
    : Node(name),
      left_wheel_prev_pos_(0.0),
      right_wheel_prev_pos_(0.0),
      x_(0.0),
      y_(0.0),
      theta_(0.0)
{

    declare_parameter("wheel_radius", 0.033);
    declare_parameter("wheel_separation", 0.17);

    wheel_radius_ = get_parameter("wheel_radius").as_double();
    wheel_separation_ = get_parameter("wheel_separation").as_double();

    RCLCPP_INFO_STREAM(get_logger(), "Wheel radius: " << wheel_radius_);
    RCLCPP_INFO_STREAM(get_logger(), "Wheel separation: " << wheel_separation_);
    prev_time_ = get_clock()->now();

    joint_sub_ = create_subscription<sensor_msgs::msg::JointState>("/joint_states", 10, std::bind(&NoisyController::joint_callback, this, _1));
    odom_pub_ = create_publisher<nav_msgs::msg::Odometry>("/bumperbot/odom_noisy", 10);

    odom_msg_.header.frame_id = "odom";
    odom_msg_.child_frame_id = "base_footprint_ekf";
    odom_msg_.pose.pose.orientation.x = 0.0;
    odom_msg_.pose.pose.orientation.y = 0.0;
    odom_msg_.pose.pose.orientation.z = 0.0;
    odom_msg_.pose.pose.orientation.w = 1.0;

    tf_br_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
    tf_stamped_msg_.header.frame_id = "odom";
    tf_stamped_msg_.child_frame_id = "base_footprint_noisy";
}

void NoisyController::joint_callback(const sensor_msgs::msg::JointState &msg)
{
    auto seed = std::chrono::system_clock::now().time_since_epoch().count();
    std::default_random_engine noise_generator(seed);
    std::normal_distribution<double> left_encoder_noise(0.0, 0.005);
    std::normal_distribution<double> right_encoder_noise(0.0, 0.005);
    double wheel_encoder_left = msg.position.at(1) + left_encoder_noise(noise_generator);
    double wheel_encoder_right = msg.position.at(0) + right_encoder_noise(noise_generator);

    // wheel encoder values
    double dp_left = wheel_encoder_left - left_wheel_prev_pos_;
    double dp_right = wheel_encoder_right - right_wheel_prev_pos_;

    rclcpp::Time msg_time = msg.header.stamp;
    rclcpp::Duration dt = msg_time - prev_time_;

    // update
    left_wheel_prev_pos_ = msg.position.at(1);
    right_wheel_prev_pos_ = msg.position.at(0);
    prev_time_ = msg_time;

    double fi_left = dp_left / dt.seconds();
    double fi_right = dp_right / dt.seconds();

    double linear_vel = (wheel_radius_ * fi_right + wheel_radius_ * fi_left) / 2.0;
    double angular_vel = (wheel_radius_ * fi_right - wheel_radius_ * fi_left) / wheel_separation_;

    double d_s = (wheel_radius_ * dp_right + wheel_radius_ * dp_left) / 2.0;
    double d_theta = (wheel_radius_ * dp_right - wheel_radius_ * dp_left) / wheel_separation_;

    // Absolute position and orientation
    theta_ += d_theta;
    x_ += d_s * cos(theta_);
    y_ += d_s * sin(theta_);

    tf2::Quaternion q;
    q.setRPY(0, 0, theta_);
    odom_msg_.pose.pose.orientation.x = q.x();
    odom_msg_.pose.pose.orientation.y = q.y();
    odom_msg_.pose.pose.orientation.z = q.z();
    odom_msg_.pose.pose.orientation.w = q.w();
    odom_msg_.pose.pose.position.x = x_;
    odom_msg_.pose.pose.position.y = y_;
    odom_msg_.twist.twist.linear.x = linear_vel;
    odom_msg_.twist.twist.angular.z = angular_vel;
    odom_msg_.header.stamp = get_clock()->now();

    // tf2 stuff
    tf_stamped_msg_.transform.translation.x = x_;
    tf_stamped_msg_.transform.translation.y = y_;
    tf_stamped_msg_.transform.rotation.x = q.x();
    tf_stamped_msg_.transform.rotation.y = q.y();
    tf_stamped_msg_.transform.rotation.z = q.z();
    tf_stamped_msg_.transform.rotation.w = q.w();
    tf_stamped_msg_.header.stamp = get_clock()->now();

    odom_pub_->publish(odom_msg_);
    tf_br_->sendTransform(tf_stamped_msg_);

    RCLCPP_INFO_STREAM(get_logger(), "linear velocity: " << linear_vel << "Angular velocity: " << angular_vel);
    RCLCPP_INFO_STREAM(get_logger(), "x: " << x_ << "y: " << y_ << "theta: " << theta_);
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NoisyController>("noisy_controller");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}